src/System.Text.Json/src/System/Text/Json/JsonUtf8Reader.cs

```diff
@@ -0,0 +1,1274 @@
+            _readerOptions = _readerOptions,
+            _stack = _stack,
+            _sequencePosition = Position,
+        };
+
+        /// <summary>
+        /// Constructs a new <see cref="JsonUtf8Reader"/> instance.
+        /// </summary>
+        /// <param name="jsonData">The ReadOnlySpan&lt;byte&gt; containing the UTF-8 encoded JSON text to process.</param>
+        /// <param name="isFinalBlock">True when the input span contains the entire data to process.
+        /// Set to false only if it is known that the input span contains partial data with more data to follow.</param>
+        /// <param name="state">If this is the first call to the ctor, pass in a default state. Otherwise,
+        /// capture the state from the previous instance of the <see cref="JsonUtf8Reader"/> and pass that back.</param>
+        /// <remarks>
+        /// Since this type is a ref struct, it is a stack-only type and all the limitations of ref structs apply to it.
+        /// This is the reason why the ctor accepts a <see cref="JsonReaderState"/>.
+        /// </remarks>
+        public JsonUtf8Reader(ReadOnlySpan<byte> jsonData, bool isFinalBlock, JsonReaderState state)
+        {
+            _buffer = jsonData;
+
+            _isFinalBlock = isFinalBlock;
+
+            // Note: We do not retain _bytesConsumed or _sequencePosition as they reset with the new input data
+            _stackFreeContainer = state._stackFreeContainer;
+            _lineNumber = state._lineNumber;
+            _lineBytePosition = state._lineBytePosition;
+            _currentDepth = state._currentDepth;
+            _maxDepth = state._maxDepth == 0 ? JsonReaderState.StackFreeMaxDepth : state._maxDepth;
+            _inObject = state._inObject;
+            _isNotPrimitive = state._isNotPrimitive;
+            _tokenType = state._tokenType;
+            _readerOptions = state._readerOptions;
+            _stack = state._stack;
+
+            _consumed = 0;
+            _totalConsumed = 0;
+            _isLastSegment = _isFinalBlock;
+            _isSingleSegment = true;
+            _currentPosition = default;
+            //_nextPosition = default;
+            _sequence = default;
+
+            IsValueMultiSegment = false;
+            ValueSpan = ReadOnlySpan<byte>.Empty;
+            ValueSequence = ReadOnlySequence<byte>.Empty;
+        }
+
+        /// <summary>
+        /// Read the next JSON token from input source.
+        /// </summary>
+        /// <returns>True if the token was read successfully, else false.</returns>
+        /// <exception cref="JsonReaderException">
+        /// Thrown when an invalid JSON token is encountered according to the JSON RFC
+        /// or if the current depth exceeds the recursive limit set by the max depth.
+        /// </exception>
+        public bool Read() => ReadSingleSegment();
+
+        private void StartObject()
+        {
+            _currentDepth++;
+            if (_currentDepth > _maxDepth)
```

Comment:
_maxDepth seems to combine two things:
- An optimization to avoid allocating a Stack
- A safety precaution against malicious/poorly formed/unexpected input data.

The latter makes sense.  Why do we need to tie the former to it, though?  e.g. why not start off without a stack, and then lazily allocate one and transfer data to it if/when we hit the limit?  Seems like it shouldn't add measurable cost, and it would allow usage to take advantage of the optimization unless it actually tripped over the internal 64 depth limit, making it pay-for-play.  Then the exposed limit would purely be about risk management.

Reply:
Aside from providing the ability to set a `MaxDepth` (as you mentioned for risk management), we also set the initial `MaxDepth` to a constant (constrain by default for safety). What's the harm in leveraging that constant value as an optimization as well? I believe the current implementation is still pay-for-play. Today, we allocate the stack up-front **once** if the max depth is set above the optimized limit (and only use it once we hit the limit). There is a downside to lazily allocating the stack at the point of hitting the limit. The check and allocation would then have to happen in a method like `StartObject` (which gets called a lot) rather than once in something like a ctor of `JsonReaderState`. And now `StartObject` is more expensive too.

**Before:**
```C#
if (_currentDepth <= JsonReaderState.StackFreeMaxDepth)
    _stackFreeContainer = (_stackFreeContainer << 1) | 1;
else
    _stack.Push(JsonTokenType.StartObject);
```

**After:**
```C#
if (_currentDepth <= JsonReaderState.StackFreeMaxDepth)
    _stackFreeContainer = (_stackFreeContainer << 1) | 1;
else
{
    if (_stack == null) _stack = new Stack<JsonTokenType>();
    _stack.Push(JsonTokenType.StartObject);
}
```

Reply:
>  I want to avoid having this check all the time and increase in method body size

Isn't this path very unexpected? You could put the entire else block into a helper:
```C#
else EnsureAndPushStack(JsonTokenType.StartObject);
```
No more IL than before.

> What's the harm in leveraging that constant value as an optimization as well? 

How is setting it to a constant an optimization?

Reply:
> Isn't this path very unexpected? You could put the entire else block into a helper

Sure, I could do that instead. I want to understand the benefit first.

> why not start off without a stack, and then lazily allocate one and transfer data to it if/when we hit the limit? 

What's the advantage of decoupling the optimization of using stack only where necessary from the safety of _maxDepth?

Reply:
> What's the advantage of decoupling the optimization of using stack only where necessary from the safety of _maxDepth?

Gaining the advantage that readers which can't guarantee that they'll stay within 64 get the lower allocation and faster lookup benefits of having a document which does stay within 64.

The DOMmy model and the serializer are probably both going to have to say "no" (or very high) max depth.

Reply:
> Gaining the advantage that readers which can't guarantee that they'll stay within 64 get the lower allocation and faster lookup benefits of having a document which does stay within 64.

Even if max depth is > 64, the reader would still benefits from **faster lookup** for depths <= 64 today. And if they cross 64, they start using the stack. It's not like I say if maxDepth > 64, I am using the stack from the get go. Now sure, I end up allocating the stack for it up front, but that is only a concern if the user sets the max depth > 64 but the payload never reaches that depth. Is that the scope of the benefit then? Avoid the **allocation** in the off chance max depth is set higher than needed?

Reply:
> Yes. My assumption is that actually will be pretty common.

Alright, sounds good. Will adjust accordingly. Thanks.

Reply:
Thanks.  If it ends up that it does cause a regression on throughput for < 64 cases, let's revisit.

Reply:
> If it ends up that it does cause a regression on throughput for < 64 cases

I didn't observe any significant regression.